### PR TITLE
Allow virtual objects to be created in batches

### PR DESCRIPTION
**Pull request by mjgiarlo**
Connects to #375 
Connects to sul-dlss/argo#1463

Includes:
* Mark `ObjectsController#update` as soon as be removed
* Make clear in `ItemQueryService` which parents are associated with which combination errors
* Add class method for convenience of checking if a work is open for versioning
* Check whether item is open or openable, NOT whether it allows modification, within `ItemQueryService`
* Handle versioning concerns when adding constituents
* Extract `VirtualObjectsController` from `ObjectsController`: it functions identically to the `ObjectsController` implementation except it can create many virtual objects at a time.
* Enhance parameter validation in `VirtualObjectsController` by using dry-schema

### CODE REVIEW

File: spec/services/version_service_spec.rb [hunk 1]
```diff
@@ -18,6 +18,20 @@
     allow(obj.inner_object).to receive(:repository).and_return(double('frepo').as_null_object)
   end
 
+  describe '.open?' do
+    let(:instance) { instance_double(described_class) }
```
Comment (jcoyne) [inline spec/services/version_service_spec.rb]: You can add `open_for_versioning?: true` here and then remove the `allow` on line 29.

File: app/controllers/virtual_objects_controller.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+class VirtualObjectsController < ApplicationController
+  # Create one or more virtual objects represented by JSON:
+  # {
+  #   virtual_objects: [
+  #     { parent_id: '', child_ids: [] }
+  #   ]
+  # }
+  def create
+    # validate that the virtual_objects parameter is an present, raises ActionController::ParameterMissing
+    params.require(:virtual_objects)
```
Comment (mjgiarlo) [inline app/controllers/virtual_objects_controller.rb]: Extract to private method, `create_params`.

File: app/controllers/virtual_objects_controller.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    filtered_params = params.permit(virtual_objects: [:parent_id, child_ids: []])
+    raise ActionController::ParameterMissing, 'virtual_objects must be an array' unless filtered_params[:virtual_objects]
```
Comment (mjgiarlo) [inline app/controllers/virtual_objects_controller.rb]: Look into parameter validation.

File: spec/services/version_service_spec.rb (hunk 1 above)
Comment (mjgiarlo) [reply spec/services/version_service_spec.rb]: Will do.

File: app/controllers/virtual_objects_controller.rb (hunk 3 above)
Comment (mjgiarlo) [reply app/controllers/virtual_objects_controller.rb]: Use Json schema?

File: app/services/constituent_service.rb [hunk 4]
```diff
@@ -18,16 +18,21 @@ def initialize(parent_druid:)
   # @raises ActiveFedora::RecordInvalid if AF object validations fail on #save!
   # @returns [NilClass, Hash] true if successful, hash of errors otherwise (if combinable validation fails)
   def add(child_druids:)
-    errors = ItemQueryService.validate_combinable_items([parent_druid] + child_druids)
+    errors = ItemQueryService.validate_combinable_items(parent: parent_druid, children: child_druids)
```
Comment (jcoyne) [inline app/services/constituent_service.rb]: 👏

File: app/services/constituent_service.rb [hunk 5]
```diff
@@ -38,11 +43,16 @@ def add(child_druids:)
 
   def add_constituent(child_druid:)
     child = ItemQueryService.find_combinable_item(child_druid)
+    # Make sure the child is open before making modifications
+    VersionService.open(child) unless VersionService.open?(child)
```
Comment (ndushay) [inline app/services/constituent_service.rb]: isn't this check now superfluous, given your changes to `find_combinable_item` ?

File: config/routes.rb [hunk 6]
```diff
@@ -20,6 +20,9 @@
       get 'catkey', to: 'marcxml#catkey'
     end
 
+    resources :virtual_objects, only: [:create]
+
+    # TODO: Remove :update once Argo, in stage and prod, uses a version of dor-services-client that no longer hits this endpoint
```
Comment (ndushay) [inline config/routes.rb]: is the migration / removal of `:update` ticketed?  If not ... make it so??

File: app/services/constituent_service.rb (hunk 5 above)
Comment (mjgiarlo) [reply app/services/constituent_service.rb]: @ndushay Nope! The changes to `ItemQueryService#find_combinable_item` make sure that every item is either open or openable before proceeding with any operation. This step makes sure every item is *open* before making any modifications. (This prevents users from having to manually open every parent and child object before beginning the virtual object creation.)

File: config/routes.rb (hunk 6 above)
Comment (mjgiarlo) [reply config/routes.rb]: I'm planning to do this immediately after sul-dlss/argo#1463. I wasn't planning to ticket every subtask bc my hope was to push through the PRs rapid-fire style so we can finish this feature, test it, and tidy things before the week is through. I am happy to write 'em up if you think it's worth the time.

File: app/services/constituent_service.rb (hunk 5 above)
Comment (ndushay) [reply app/services/constituent_service.rb]: ah ha!  Cool!

File: config/routes.rb (hunk 6 above)
Comment (ndushay) [reply config/routes.rb]: your plan makes sense.  :-)

**Pull request merged**